crypto/ec/ec_lib.c

```diff
@@ -328,6 +336,7 @@ int EC_GROUP_set_generator(EC_GROUP *group, const EC_POINT *generator,
     }
 
     BN_MONT_CTX_free(group->mont_data);
+ end:
```

Comment:
What is the rationale behind clearing `group->mont_data` but not freeing it? Isn't it cleared in `EC_GROUP_new()` and never touched again (in FIPS mode)?

Reply:
Upon debugging I realized that for some reason using the FIPS module that field is basically uninitialized: it's not NULL but it's not even a valid pointer.

Upon further analysis: the `EC_GROUP_new()` function in `OpenSSL_1_0_2-stable` does a malloc and manually sets to `NULL` most optional pointers, including `group->mont_data`.
But, the `EC_GROUP_new()` function in the FIPS module does not even mention the `mont_data` field which stays uninitialized.

On a related note, answering to @richsalz (https://github.com/openssl/openssl/pull/7135#issuecomment-419141891) it looks like the FIPS module replaces only a subset of the symbols of libcrypto.
It looks like `EC_GROUP_new()` implementation comes from the FIPS module for example, but `EC_GROUP_set_generator()` is coming from libcrypto outside of the FIPS module.

This is starting to worry me, and I'm trying to dig deeper, especially I am trying to understand how any `*_new()` method is working when in FIPS mode...
What's worrying is that, until now, to me it looks like, in FIPS mode:
- `EC_GROUP_new()` comes from the FIPS module, which is compiled with a set of headers which has been set in stone, which includes a definition of `struct ec_group_st` (a.k.a. `EC_GROUP`) which is apparently used to determine the size of an `EC_GROUP` object for the `malloc` in `EC_GROUP_new()` inside the FIPS module;
- some of the functions manipulating an `EC_GROUP` pointer come from outside the FIPS module (e.g., in this case `EC_GROUP_set_generator()`), which use the `struct ec_group_st` definition included in the `OpenSSL_1_0_2-stable` branch, which differs from the one in the FIPS module (e.g., the `mont_data` member is defined in the stable branch, but is not defined in the FIPS module headers)

This looks like a recipe for potential OOB reads/writes on top of the issue of uninitialized fields: I would expect random segfaults to pop out immediately, but I'm not seeing this, so I am trying to understand what kind of magic glue is keeping everything from falling apart....

Reply:
> It looks like EC_GROUP_new() implementation comes from the FIPS module for example,
> but EC_GROUP_set_generator() is coming from libcrypto outside of the FIPS module.

I would not only be worried about OOB r/w (of course it's important), but also about the fact that mixing FIPS-validated and non-FIPS-validated crypto algorithms is imho a clear violation of the FIPS rules.

Reply:
I would expect fipssyms.h to remap EC_GROUP_set_generator()...

Reply:
To the point @romen raises, it does seem quite odd to be jumping in and out of the FIPS module while operating on a single object. Perhaps the core issue here is something as simple a failure to include fipssyms.h somewhere, and thereby failing to redirect call into the module? Or perhaps some methods aren't defined in fipssyms.h that should be? I realize the FOM itself can't change. But my understanding is that the header itself is outside the module. If an operation isn't supported in the FOM, it seems far more sensible to have some kind of failure - preferably at build-time - than to suddenly redirect outside of the FOM for an object that up to that point the FOM had been managing.

Reply:
As @jwalch presumed, fipssyms.h contains

```#define EC_GROUP_set_generator FIPS_ec_group_set_generator```

so if the header is visible, then actually it should be easy to verify in the debugger that the latter function is entered and not the former.

>Upon debugging I realized that for some reason using the FIPS module that field is basically  uninitialized: it's not NULL but it's not even a valid pointer.

One possible explanation for that observation would be that you viewed the contents of the structure inside gdb from a frame which only sees the 1.0.2 headers but not the FIPS object module headers? Then it would only be a debugger display problem. Also this question could be answered  by stepping into `FIPS_ec_group_set_generator`.

Reply:
If `group` was indeed allocated by the FOM, then it is probably not a good idea to clear the non-existing `mont_data` member.

Reply:
I just checked the headers: the `mont_data` member was added in 8aed2a75483 and is not present in the FOM.

Reply:
Right! This brings some memories back... There is EC_GROUP_VERSION macro, which allows to tell apart `group`-s allocated by FIPS module and 1.0.2 libcrypto. In other words one needs to use EC_GROUP_VERSION to guard mont_data filed usage. It should be possible to replace check for FIPS_mode as `if (EC_GROUP_VERSION(group) && BN_is_odd(&group->order)`. Of course  ec_precompute_mont_data opens with EC_GROUP_VERSION check, so that one would be redundant. [My availability is limited for few days, and it's easier to comment than to create request :-( And singing off for today...]

Reply:
On further analysis, it looks like the magic glue is in the form of a mix of preprocessor macros that rename some OpenSSL function names into fips specific names (see `fips/fipssyms.h` in the FOM source) and a PERL script that does some kind of magic I don't completely understand  (see the contents of `fips/tools` in the FOM source).

Inside the FIPS module any reference to `EC_GROUP_new()` is actually an alias for `fips_ec_group_new()` (which doesn't know anything about fields like `group->mont_data`).
Outside the FIPS module, any reference to `EC_GROUP_new()` will actually call the `EC_GROUP_new()` defined outside the FOM.

Referencing the sample code above, when
```c
    EC_GROUP *ecgroup = EC_GROUP_new_by_curve_name(NID_secp224k1);
```
is called, from libcrypto in FIPS mode, the code in libcrypto calls `fips_ec_group_new_by_curve_name` (exported by the FOM), so the returned `EC_GROUP` pointer will point to a `EC_GROUP` structure as defined inside the FOM (i.e., shorter than a libcrypto `EC_GROUP` object, as it does not include a `mont_data` member): let's call the object returned by reference a `FIPS_EC_GROUP` object.

When
```c
    TS_ASSERT_EQUALS(1, EC_KEY_set_group(ecdsa_key, ecgroup));
```
is called, libcrypto's `EC_GROUP_dup()` is called to assign a duplicate of the `FIPS_EC_GROUP* ecgroup` to the `EC_KEY* ecdsa_key`: libcrypto's `EC_GROUP_dup()` allocates a new object using libcrypto's `EC_GROUP_new()`, and then calls libcrypto's `EC_GROUP_copy()` to copy the contents of the `FIPS_EC_GROUP` source object into the newly created `EC_GROUP` dest object.

libcrypto's `EC_GROUP_copy()` uses the `EC_GROUP_VERSION()` macro to guard against reading from a nonexistent `mont_data` field of a src `FIPS_EC_GROUP` object, or from freeing/setting a nonexistent `mont_data` field of a dest `FIPS_EC_GROUP`.

When
```c
    EC_KEY *private_key = PEM_read_bio_ECPrivateKey(bp_private, NULL, NULL, NULL);
```
is called, this will happen:
```
#0  EC_GROUP_new_curve_GFp (p=0x555555777db0, a=0x555555777d30, b=0x555555777d70, ctx=0x0) at ec_cvt.c:86
#1  0x00007ffff7a2a473 in ec_asn1_parameters2group (params=0x555555774140) at ec_asn1.c:844
#2  0x00007ffff7a2a96f in ec_asn1_pkparameters2group (params=0x5555557747c0) at ec_asn1.c:956
#3  0x00007ffff7a2aa96 in d2i_ECPKParameters (a=0x555555774108, in=0x7fffffffd5d0, len=149) at ec_asn1.c:986
#4  0x00007ffff7a2b647 in d2i_ECParameters (a=0x0, in=0x7fffffffd5d0, len=149) at ec_asn1.c:1255
#5  0x00007ffff7a2f6da in eckey_type2param (ptype=16, pval=0x5555557740c0) at ec_ameth.c:153
#6  0x00007ffff7a2fa81 in eckey_priv_decode (pkey=0x5555557747e0, p8=0x555555777560) at ec_ameth.c:249
#7  0x00007ffff7a80936 in EVP_PKCS82PKEY (p8=0x555555777560) at evp_pkey.c:91
#8  0x00007ffff7ab6c59 in PEM_read_bio_PrivateKey (bp=0x5555557732b0, x=0x0, cb=0x0, u=0x0) at pem_pkey.c:97
#9  0x00007ffff7ab5402 in PEM_read_bio_ECPrivateKey (bp=0x5555557732b0, key=0x0, cb=0x0, u=0x0) at pem_all.c:354
#10 0x0000555555554ed0 in main () at issue7127.c:52
```
`EC_GROUP_new_curve_GFp` is from libcrypto, but in FIPS mode it will execute
```c
    if (FIPS_mode())
        return FIPS_ec_group_new_curve_gfp(p, a, b, ctx);
```
, so the returned pointer will reference a `FIPS_EC_GROUP` object.

Later, in `EC_GROUP_set_generator()`, we are unconditionally accessing the `mont_data` field, but any access to members of `EC_GROUP` which are not defined in the `FIPS_EC_GROUP` definitions should always be guarded by an `EC_GROUP_VERSION()` check.

I'll update the PR shortly with the results of this analysis.

Reply:
I see only now that you all went ahead while I was writing the above report and actually came to the same conclusions. :)

Reply:
Nice writeup, though ;-).